server/src/main/java/org/elasticsearch/health/node/DiskHealthIndicatorService.java

```diff
@@ -0,0 +1,284 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.health.node;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.elasticsearch.cluster.coordination.StableMasterHealthIndicatorService;
+import org.elasticsearch.cluster.metadata.IndexMetadata;
+import org.elasticsearch.cluster.node.DiscoveryNode;
+import org.elasticsearch.cluster.node.DiscoveryNodeRole;
+import org.elasticsearch.cluster.service.ClusterService;
+import org.elasticsearch.health.Diagnosis;
+import org.elasticsearch.health.HealthIndicatorDetails;
+import org.elasticsearch.health.HealthIndicatorImpact;
+import org.elasticsearch.health.HealthIndicatorResult;
+import org.elasticsearch.health.HealthIndicatorService;
+import org.elasticsearch.health.HealthStatus;
+import org.elasticsearch.health.ImpactArea;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.Locale;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+public class DiskHealthIndicatorService implements HealthIndicatorService {
+    public static final String NAME = "disk";
+
+    private static final Logger logger = LogManager.getLogger(DiskHealthIndicatorService.class);
+
+    private final ClusterService clusterService;
+
+    private static final HealthIndicatorImpact RED_HEALTH_AFFECTED_INDICES_INGEST_IMPACT = new HealthIndicatorImpact(
+        1,
+        "Cannot insert or update documents in the affected indices.",
+        List.of(ImpactArea.INGEST)
+    );
+
+    private static final HealthIndicatorImpact YELLOW_HEALTH_AFFECTED_INDICES_INGEST_IMPACT = new HealthIndicatorImpact(
+        1,
+        "May not be able to insert or update documents in the affected indices.",
+        List.of(ImpactArea.INGEST)
+    );
+
+    private static final HealthIndicatorImpact HEALTH_INGEST_NODES_PIPELINES_IMPACT = new HealthIndicatorImpact(
+        2,
+        "Cannot run ingest pipelines",
```

Comment:
I am assuming that this is capturing the case when a node has the ingest role and no data, right? Have we verified that running low on disk space is causing the ingest pipelines to fail? If we haven't verified it I am wondering if this is worth adding already?

Reply:
This is independent of the data role (it could have both). But you are right that I have not tested any of these impacts. My assumption is that if an ingest node has a bad disk then pipelines are at serious risk of not running on that box though.